table/tables/state_remote.go

```diff
@@ -276,3 +297,232 @@ func (r *mockStateRemoteData) LockForWrite(tid int64, now, ts uint64) (uint64, e
+		}
+
+		switch lockType {
+		case CachedTableLockNone:
+		case CachedTableLockRead:
+		case CachedTableLockWrite, CachedTableLockIntend:
+			return nil
+		}
+		succ = true
+		if err := h.updateRow(ctx, tid, "READ", ts); err != nil {
+			return errors.Trace(err)
+		}
+
+		return nil
+	})
+	return succ, err
+}
+
+func (h *stateRemoteHandle) LockForWrite(ctx context.Context, tid int64, now, ts uint64) error {
+	h.Lock()
+	defer h.Unlock()
+	for {
+		retry, err := h.lockForWriteOnce(ctx, tid, now, ts)
+		if err != nil {
+			return err
+		}
+		if !retry {
+			break
+		}
+
+		store := h.exec.GetStore()
+		o := store.GetOracle()
+		newTS, err := o.GetTimestamp(ctx, &oracle.Option{TxnScope: kv.GlobalTxnScope})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		now, ts = newTS, leaseFromTS(newTS)
+	}
+	return nil
+}
+
+func (h *stateRemoteHandle) lockForWriteOnce(ctx context.Context, tid int64, now, ts uint64) ( /*retry*/ bool, error) {
+	err := h.beginTxn(ctx)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	defer func() {
+		if err != nil {
+			terror.Log(h.rollbackTxn(ctx))
+		}
+	}()
+
+	lockType, lease, oldReadLease, err := h.loadRow(ctx, tid)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	// The lease is outdated, so lock is invalid, clear orphan lock of any kind.
+	if now > lease {
+		if err := h.updateRow(ctx, tid, "WRITE", ts); err != nil {
+			return false, errors.Trace(err)
+		}
+		return false, h.commitTxn(ctx)
+	}
+
+	// The lease is valid.
+	switch lockType {
+	case CachedTableLockNone:
+		if err = h.updateRow(ctx, tid, "WRITE", ts); err != nil {
+			return false, errors.Trace(err)
+		}
+		return false, h.commitTxn(ctx)
+	case CachedTableLockRead:
+		// Change from READ to INTEND
+		if _, err = h.execSQL(ctx, "update mysql.table_cache_meta set lock_type='INTEND', oldReadLease=%?, lease=%? where tid=%?", lease, ts, tid); err != nil {
+			return false, errors.Trace(err)
+		}
+		if err = h.commitTxn(ctx); err != nil {
+			return false, errors.Trace(err)
+		}
+
+		// Wait for lease to expire, and then retry.
+		waitForLeaseExpire(oldReadLease, now)
+		return true, nil
+	case CachedTableLockIntend, CachedTableLockWrite:
+		// `now` exceed `oldReadLease` means wait for READ lock lease is done, it's safe to read here.
+		if now > oldReadLease {
+			if lockType == CachedTableLockIntend {
+				if err = h.updateRow(ctx, tid, "WRITE", ts); err != nil {
+					return false, errors.Trace(err)
+				}
+			}
+			return false, h.commitTxn(ctx)
+		}
+
+		// Otherwise, the WRITE should wait for the READ lease expire.
+		terror.Log(h.rollbackTxn(ctx))
+		waitForLeaseExpire(oldReadLease, now)
+		// And then retry change the lock to WRITE
+		return true, nil
+	}
+	return false, errors.New("should never run here")
+}
+
+func waitForLeaseExpire(oldReadLease, now uint64) {
+	if oldReadLease >= now {
+		t1 := oracle.GetTimeFromTS(oldReadLease)
+		t2 := oracle.GetTimeFromTS(now)
+		waitDuration := t1.Sub(t2)
+		time.Sleep(waitDuration)
```

Comment:
It'll will be a long time if now is very small... for example

```
begin;
do sleep(3600);
insert into t values(1); -- LockForWrite, startTS (now) is 1 hour ago
```

Reply:
May be we should use now as current ts instead of startTS

Reply:
Yes, now should be current ts instead of startTS ...
But the cost is a extra network round trip to PD for the current ts

Maybe we can compare the clock time now() with start ts, and if it's differ too much, we can ask PD for current ts

This can be done in the caller of `LockForWrite`